src/org/nfrac/comportex/cells.cljc

```diff
@@ -370,16 +371,23 @@
   the proximal excitation but is given keyed by cell rather than by
   column. Normally only cells with some proximal input are included,
   but if `spontaneous-activation?` is true, this is relaxed
-  (i.e. prediction alone could cause activation)."
+  (i.e. prediction alone could cause activation).
+
+  * col-exc is keyed by column as [col 0].
+  * tp-exc is keyed by cell as [col ci]."
   [col-exc tp-exc distal-exc distal-weight spontaneous-activation? depth]
-  (let [;; add TP columns to list of columns to consider (may have no proximal)
-        basic-col-exc (merge-with + col-exc
-                                  (into {} (map (fn [[col _]] [col 0])) (keys tp-exc)))
+  (let [has-tp? (seq tp-exc)
+        ;; add TP columns to list of columns to consider (may have no proximal)
+        basic-col-exc (if has-tp?
+                        (merge-with + col-exc
+                                    (into {} (map (fn [[col _]] [[col 0] 0]))
+                                          (keys tp-exc)))
```

Comment:
Since we're just adding zero, I think this would be easier to read as a simple merge, putting the zeros on the left so that they get overwritten.

e.g.

~~~clojure
(merge (zipmap (for [[col _] (keys tp-exc)]
                 [col 0])
               (repeat 0))
       col-exc)
~~~

and wrapping it in `has-tp?` is optional, as it is now.